Documentation for End to End Testing (#12459)
forem/forem: docs/tests/e2e-tests.md

---
title: End to End Tests
---

# End to End Tests

[Cypress](https://www.cypress.io), a JavaScript based framework, is used for end
to end (E2E) testing. Tests and associated utilities are within the `cypress`
directory.

```shell
$ tree app/javascript/article-form -L 1
cypress
├── fixtures
|   └── initialAdmin.json
|   └── ...
├── integration
|   └── **/*.spec.js
├── plugins

> └── support

Reviewer: Does this need updated? The `tree` command is pointing to the `article-form` dir rather than the cypress one, and the fixtures/integration dirs are now organised with sub directories

Author: Done. I just ran the command `tree cypress` to get the actual output instead of hacking a pre existing one. 🙃

```

In addition to Cypress, we use
[cypress-testing-library](https://github.com/testing-library/cypress-testing-library).
This provides custom Cypress commands and utilities to improve how E2E tests are
written. dom-testing-library and preact-testing-library are already used for

> [front-end tests](frontend-tests.md), which both offer a similar API in the

Reviewer: Wee change needed in this link
```suggestion
[front-end tests](/tests/frontend-tests), which both offer a similar API in the
```

context of testing a Preact or Stimulus component. All these tools are part of
the [Testing Library](https://testing-library.com) family.

For the test web server, the
[cypress-rails](https://github.com/testdouble/cypress-rails) gem is used to

> start a test web server that runs a rails test environemnt (`RAILS_ENV=test`).

Reviewer: ```suggestion
start a test web server that runs a rails test environment (`RAILS_ENV=test`).
```

> It also resets the databaase between test runs by starting a database

Reviewer: ```suggestion
It also resets the database between test runs by starting a database
```

> transaction at the beginning of a test and perfors a rollback at the end of a

Reviewer: ```suggestion
transaction at the beginning of a test and performs a rollback at the end of a
```

test being run. The cypress-rails gem also provides a rake task that allows us
to coordinate all this work.

## Running E2E Tests Locally

From the command line, run `yarn e2e` to start E2E testing.

1. `bundle check` will run to ensure all the gems for the project are up to
   date. If they are not, you will be prompted to run `bundle install`.

```bash
Doing a quick bundle check to make sure gems are all up to date.

 * zonebie (0.6.1)
Install missing gems with `bundle install`
Unable to launch end to end tests. Ensure that all your gems are installed and up to date.
```

2. `yarn install` will run to ensure front-end packages are up to date.

```bash
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
success Already up-to-date.
```

> 3. You will be prompted to set up the end to end (E2E) test database. Type `y`
>    or `Y` to install the E2E test database. Typically this is a one time thing,
>    but it can be run as well if ever you corrupt your database.

Reviewer: I'm wondering if we can make it more clear the prompt will appear every time, but they can (in most cases) safely select `n` after the first time - something like:

```suggestion
3. You will be prompted to set up the end to end (E2E) test database. Type `y`
   or `Y` to install the E2E test database. Typically you only need to select `y` the first time you run e2e tests,
   but it can also be run if ever you corrupt your database and need to reset it back to its original state.
```

```bash
Do you need to set up your end to end (E2E) testing database? Answer yes
if this is your first time running E2E tests on your local machine or you
need to recreate your E2E test database. (y/n)
```

4. If `y` or `Y` is pressed, the E2E test database will install.

```bash
Setting up the E2E database before running E2E tests...

Dropped database 'PracticalDeveloper_test'
Created database 'PracticalDeveloper_test'

...
```

> 5. The Cypress test runner will open and you are now ready to run end to end
>    tests.

Reviewer: Thought it might be useful to add a link here, in case people want to see a short walkthrough of the test runner controls/features

```suggestion
5. The [Cypress test runner](https://docs.cypress.io/guides/core-concepts/test-runner.html#Overview) will open and you are now ready to run end to end
   tests.
```

---
title: End to End Tests
---

# End to End Tests

[Cypress](https://www.cypress.io), a JavaScript based framework, is used for end
to end (E2E) testing. Tests and associated utilities are within the `cypress`
directory.

```shell
$ tree cypress

cypress
├── fixtures
│   ├── images
│   │   └── admin-image.png
│   └── users
│       ├── adminUser.json
│       └── changePasswordUser.json
├── integration
│   ├── adminFlows
│   │   └── config
│   │       └── authenticationSection.spec.js
│   └── loginFlows
│       ├── userChangePassword.spec.js
│       └── userLogin.spec.js
├── plugins
│   └── index.js
└── support
    ├── commands.js
    └── index.js
```

In addition to Cypress, we use
[cypress-testing-library](https://github.com/testing-library/cypress-testing-library).
This provides custom Cypress commands and utilities to improve how E2E tests are

> written. dom-testing-library and preact-testing-library are already used for

Reviewer 2: ```suggestion
written. `dom-testing-library` and `preact-testing-library` are already used for
```

> [front-end tests](/tests/frontend-tests), which both offer a similar API in the

Reviewer: I always always forget which one we're using here 😅

context of testing a Preact or Stimulus component. All these tools are part of
the [Testing Library](https://testing-library.com) family.

For the test web server, the
[cypress-rails](https://github.com/testdouble/cypress-rails) gem is used to
start a test web server that runs a rails test environment (`RAILS_ENV=test`).
It also resets the database between test runs by starting a database transaction
at the beginning of a test and performs a rollback at the end of a test being

> run. The cypress-rails gem also provides a rake task that allows us to

Reviewer 2: ```suggestion
run. The `cypress-rails` gem also provides a rake task that allows us to
```
